src/generators.jl

```diff
@@ -802,18 +802,159 @@ end
+- `seed::Union{Nothing,UInt}' - An optional seed for the network.
+- `random_node'::Int -> Int  - The function to randomly select parents. 
+Reference: 
+
+    (section 4.2.1) Graph Evolution: Densification and Shrinking Diameters
+    Jure Leskovec, Jon Kleinberg, Christos Faloutsos
+    https://arxiv.org/pdf/physics/0603229.pdf
+
+Functions
+---------
+* `forest_fire_graph(n)` Generate an n node graph from a 10 node clique with a .4 
+   burn probability. 
+* `forest_fire_graph(G,n,p)` Generate an n node graph from the seed graph G, with a 
+   burn probability of p. 
+* `forest_fire_graph(::Type{S},G,n,p)` Generate an n node graph from the seed 
+   graph G, with a probability of p for vertex labels of type S. 
+
+Output
+------
+- `A` - the newly generated network of the same type as `seed_network`.
+- 'parents::Array{Int,1}': The list of vertices selected to start the random
+   walks from. Vertices who are their own parents are from the seed_network.
+""" 
+function forest_fire_graph(target_size::Int,m::Int=10, 
+    burn_p::Float64=.4; kwargs...)
+    clique = sprand(m,m,1.0)
+    clique[1:(m+1):m^2] .= 0
+    dropzeros!(clique)
+    return forest_fire_graph(eltype(clique.rowval),clique,target_size,burn_p;kwargs...)
+end 
+
+forest_fire_graph(seed_network::Union{MatrixNetwork,SparseMatrixCSC},target_size::Int, burn_p::Float64;kwargs...) = 
+                forest_fire_graph(eltype(seed_network.rowval),seed_network,target_size, burn_p; kwargs...)
+
+function forest_fire_graph(::Type{S},seed_network::Union{MatrixNetwork,SparseMatrixCSC}, 
+                           target_size::Int, burn_p::Float64;
+                           seed::Union{Nothing,UInt}=nothing,
+                           random_node=x->rand(1:x)) where S
+
+    if !(seed === nothing)
+        seed!(seed)
+    end 
+
+    graph_size = size(seed_network,1)
+    steps = target_size - graph_size
+    parents = Vector{S}(undef,target_size)
+    parents[1:graph_size] .= 1:graph_size
+
+    toburn = Set{S}()
+    burnt = Set{S}()
+    burning = Set{S}()
+
+    alive = Vector{S}(undef,0)
+
+
+    #  -- convert seed_graph to List of Lists format -- #
+    neighbors = matrix_to_list_of_list(S,seed_network) # assuming symmetric network
+    for _ = 1:steps
+        push!(neighbors,Vector{S}(undef,0))
+    end
+
+    for v_i = (graph_size+1):target_size
+        parent = random_node(graph_size)
+        
+        new_v = graph_size+1
+        _burn!(neighbors,new_v,parent,burn_p,
+                     toburn,burnt,burning,alive)
+
+        parents[v_i] = parent
+        for v_j in neighbors[new_v]
+            push!(neighbors[v_j],new_v)
+        end 
+        graph_size += 1
+
+    end 
+    
+    A =  list_of_list_to_sparse_matrix(neighbors)
+    if typeof(seed_network) === MatrixNetwork
+        A = MatrixNetwork(A)
+    end
+    return A, parents
+end 
+
+
+function _burn!(neighbors::Vector{Vector{S}},new_v::Int,v0::Int,burn_p::Float64,
```

Comment:
IF this is easy --- can we make this a top level function? e.g. there are other instances where you want to do forest fire sampling... Since the process is the same, it'd be nice to be able to reuse the function.

Reply:
What's the best interface for this to be a top function? `burn(A,v0::Int,p::Float64)` for just a single step and return final neighbor list of the random walk? Or something like `burn(A,steps::Int, rand_v::Int->Int)` which runs the process for a number of steps and selects random nodes by calling `rand_v()`.

Reply:
burn(A,v0,p) and burn!(A,v0,p,temparrays...) are the best top level interface